system/common.cpp

```diff
@@ -205,5 +211,127 @@ int WEAK pthread_cancel(struct __pthread* t)
+		uint32_t cp = str->charCodeAt(i);
+
+		if (cp >= 0xD800 && cp <= 0xDFFF) {
+			if (i + 1 < slen) {
+				uint32_t trail = str->charCodeAt(++i);
+				cp = 0x10000 + ((cp & 0x3FF) | (trail & 0x3FF));
+			} else {
+				// Missing lower surrogate
+				cp = INVALID_CODEPOINT;
+			}
+		}
+
+		if (cp > MAX_CODEPOINT)
+			cp = INVALID_CODEPOINT;
+
+		if (cp <= 0x7F) {
+			if (j < dlen)
+				*dest++ = static_cast<char>(cp);
+
+			j += 1;
+		} else if (cp <= 0x7FF) {
+			if (j + 1 < dlen) {
+				*dest++ = 0xC0 | (cp >> 6);
+				*dest++ = 0x80 | (cp & 63);
+			}
+
+			j += 2;
+		} else if (cp <= 0xFFFF) {
+			if (j + 2 < dlen) {
+				*dest++ = 0xE0 | (cp >> 12);
+				*dest++ = 0x80 | ((cp >> 6) & 63);
+				*dest++ = 0x80 | (cp & 63);
+			}
+
+			j += 3;
+		} else {
+			if (j + 3 < dlen) {
+				*dest++ = 0xF0 | (cp >> 18);
+				*dest++ = 0x80 | ((cp >> 12) & 63);
+				*dest++ = 0x80 | ((cp >> 6) & 63);
+				*dest++ = 0x80 | (cp & 63);
+			}
+
+			j += 4;
+		}
+	}
+	return j;
+}
+
+[[cheerp::genericjs]] const void *__builtin_cheerp_environ();
+[[cheerp::genericjs]] const void *__builtin_cheerp_argv();
+
+static constexpr size_t MAX_ENTRIES = 64;
+static size_t buf_size = 0;
+static char argv_environ_buf[MAX_ENTRIES * 1024];
+
+[[cheerp::genericjs]] static size_t read_to_buf(uintptr_t* entries, size_t n, const client::TArray<client::String> *arr)
+{
+	const size_t count = arr->get_length();
+	for (size_t i = 0; i < count && i < n; ++i)
+	{
+		const size_t rem = sizeof argv_environ_buf - buf_size;
+
+		const size_t len = client_to_utf8(&argv_environ_buf[buf_size], rem, (*arr)[i]);
+		if (len >= rem)
+			abort();
+
+		entries[i] = reinterpret_cast<uintptr_t>(&argv_environ_buf[buf_size]);
+
+		buf_size += len;
+		argv_environ_buf[buf_size++] = '\0';
+	}
+	return count;
+}
+
+[[cheerp::genericjs]] static size_t load_args(uintptr_t* entries, size_t n)
+{
+	return read_to_buf(entries, n, reinterpret_cast<const client::TArray<client::String>*>(__builtin_cheerp_argv()));
 }
 
+void WEAK __syscall_main_args(int* argc_p, char*** argv_p)
+{
+	static uintptr_t argv[MAX_ENTRIES];
+	const size_t count = load_args(argv, sizeof(argv)/sizeof(argv[0]));
+	if (count >= sizeof(argv)/sizeof(argv[0]))
+		abort();
+
+	*argc_p = count;
+	*argv_p = reinterpret_cast<char**>(argv);
+}
+
+[[cheerp::genericjs]] static size_t load_environ(uintptr_t* entries, size_t n)
+{
+	return read_to_buf(entries, n, reinterpret_cast<const client::TArray<client::String>*>(__builtin_cheerp_environ()));
+}
+
+extern "C" char **environ;
+void WEAK __syscall_main_environ() {
+	static uintptr_t cheerp_environ[MAX_ENTRIES];
+	const size_t count = load_environ(cheerp_environ, sizeof(cheerp_environ)/sizeof(cheerp_environ[0]));
+	if (count >= sizeof(cheerp_environ)/sizeof(cheerp_environ[0]))
+		abort();
+
+	cheerp_environ[count] = 0;
+	environ = reinterpret_cast<char**>(cheerp_environ);
+}
+#else // !__ASMJS__
+extern "C" char **environ;
+void WEAK __syscall_main_environ() {
+	environ = nullptr;
```

Comment:
any reason not to implement it for the genericjs target? Shouldn't it just work, even without any ifdef here?